<a href="https://colab.research.google.com/github/viktoruebelhart/pos_facens/blob/main/Numpy.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

 # Revisão de Estruturas de Dados em Python: Listas, Dicionários e Tuplas

Este notebook apresenta uma revisão básica das principais estruturas de dados em Python: Listas, Dicionários e Tuplas. Para cada uma delas, exploraremos seus conceitos fundamentais, exemplos de uso com funções comuns e um exercício prático.

 ## 1. Listas

### Teoria Básica

Listas em Python são coleções ordenadas e mutáveis (ou seja, podem ser alteradas) de itens. Elas podem armazenar elementos de diferentes tipos de dados (inteiros, strings, floats, outras listas, etc.). São definidas usando colchetes `[]` e os itens são separados por vírgulas.

In [ ]:
# Exemplo de uso de Listas

# Criando uma lista
minha_lista = [1, 'olá', 3.14, True]
print(f"Lista original: {minha_lista}")

# Acessando elementos (índices começam em 0)
print(f"Primeiro elemento: {minha_lista[0]}")
print(f"Último elemento: {minha_lista[-1]}")

# Modificando elementos
minha_lista[1] = 'mundo'
print(f"Lista após modificação: {minha_lista}")

# Adicionando elementos
minha_lista.append(5)
print(f"Lista após append(5): {minha_lista}")

minha_lista.insert(1, 'novo item') # Insere 'novo item' no índice 1
print(f"Lista após insert(1, 'novo item'): {minha_lista}")

# Removendo elementos
minha_lista.remove(True) # Remove a primeira ocorrência do valor True
print(f"Lista após remove(True): {minha_lista}")

item_removido = minha_lista.pop(0) # Remove e retorna o item no índice 0
print(f"Item removido com pop(0): {item_removido}")
print(f"Lista após pop(0): {minha_lista}")

# Concatenando listas
lista_extra = [10, 20]
lista_concatenada = minha_lista + lista_extra
print(f"Lista concatenada: {lista_concatenada}")

# Estendendo uma lista com outra
minha_lista.extend([6, 7])
print(f"Lista após extend([6, 7]): {minha_lista}")

# Verificando o tamanho da lista
print(f"Tamanho da lista: {len(minha_lista)}")

# Ordenando uma lista (se os elementos forem comparáveis)
numeros = [3, 1, 4, 1, 5, 9, 2]
numeros.sort()
print(f"Lista ordenada: {numeros}")

# Invertendo a ordem de uma lista
numeros.reverse()
print(f"Lista invertida: {numeros}")

# Contando ocorrências de um elemento
print(f"Número de '1' na lista: {numeros.count(1)}")

Lista original: [1, 'olá', 3.14, True]
Primeiro elemento: 1
Último elemento: True
Lista após modificação: [1, 'mundo', 3.14, True]
Lista após append(5): [1, 'mundo', 3.14, True, 5]
Lista após insert(1, 'novo item'): [1, 'novo item', 'mundo', 3.14, True, 5]
Lista após remove(True): ['novo item', 'mundo', 3.14, True, 5]
Item removido com pop(0): novo item
Lista após pop(0): ['mundo', 3.14, True, 5]
Lista concatenada: ['mundo', 3.14, True, 5, 10, 20]
Lista após extend([6, 7]): ['mundo', 3.14, True, 5, 6, 7]
Tamanho da lista: 6
Lista ordenada: [1, 1, 2, 3, 4, 5, 9]
Lista invertida: [9, 5, 4, 3, 2, 1, 1]
Número de '1' na lista: 2


 ### Exercício 1: Manipulação de Lista de Compras

Crie uma lista chamada `lista_compras` com os seguintes itens iniciais: "Maçã", "Pão", "Leite". Em seguida, realize as seguintes operações:

1.  Adicione "Ovos" ao final da lista.
2.  Insira "Queijo" na segunda posição da lista (índice 1).
3.  Remova "Pão" da lista.
4.  Imprima a lista final e o número total de itens nela.

In [13]:
# Escreva o código aqui
lista_compras = ['Maça', 'Pão', 'Leite']
lista_compras.append('Ovos')
print(f"Adicionando Ovos a Lista: {lista_compras}")
lista_compras.insert(1, 'Queijo')
print(f"Adicionando Queijo na posição 2: {lista_compras}")
lista_compras.remove('Pão')
print(f"Excluindo Pão da Lista: {lista_compras}")
print(f"Lista final: {lista_compras}, Total de itens: {len(lista_compras)}")



Adicionando Ovos a Lista: ['Maça', 'Pão', 'Leite', 'Ovos']
Adicionando Queijo na posição 2: ['Maça', 'Queijo', 'Pão', 'Leite', 'Ovos']
Excluindo Pão da Lista: ['Maça', 'Queijo', 'Leite', 'Ovos']
Lista final: ['Maça', 'Queijo', 'Leite', 'Ovos'], Total de itens: 4


 ## 2. Dicionários

### Teoria Básica

Dicionários em Python são coleções não ordenadas e mutáveis de pares `chave: valor`. Cada chave deve ser única e imutável (como strings, números ou tuplas), e os valores podem ser de qualquer tipo de dado. Dicionários são definidos usando chaves `{}` e os pares são separados por vírgulas.

In [ ]:
# Exemplo de uso de Dicionários

# Criando um dicionário
dicionario_aluno = {
    'nome': 'Ana',
    'idade': 20,
    'curso': 'Engenharia',
    'notas': [8.5, 9.0, 7.8]
}
print(f"Dicionário original: {dicionario_aluno}")

# Acessando valores
print(f"Nome do aluno: {dicionario_aluno['nome']}")
print(f"Idade do aluno (usando get): {dicionario_aluno.get('idade')}")
print(f"Curso do aluno (chave inexistente com get, valor padrão): {dicionario_aluno.get('turma', 'Não informado')}")

# Modificando valores
dicionario_aluno['idade'] = 21
print(f"Dicionário após modificar idade: {dicionario_aluno}")

# Adicionando novos pares chave:valor
dicionario_aluno['cidade'] = 'São Paulo'
print(f"Dicionário após adicionar cidade: {dicionario_aluno}")

# Removendo pares chave:valor
curso_removido = dicionario_aluno.pop('curso') # Remove e retorna o valor associado à chave
print(f"Curso removido: {curso_removido}")
print(f"Dicionário após remover curso: {dicionario_aluno}")

del dicionario_aluno['notas'] # Remove o par 'notas': [...] usando del
print(f"Dicionário após remover notas com del: {dicionario_aluno}")

# Obtendo todas as chaves
print(f"Chaves do dicionário: {dicionario_aluno.keys()}")

# Obtendo todos os valores
print(f"Valores do dicionário: {dicionario_aluno.values()}")

# Obtendo todos os itens (pares chave:valor)
print(f"Itens do dicionário: {dicionario_aluno.items()}")

# Verificando o tamanho do dicionário
print(f"Número de pares no dicionário: {len(dicionario_aluno)}")

# Atualizando um dicionário com outro
dados_complementares = {'estado': 'SP', 'matricula': '2023001'}
dicionario_aluno.update(dados_complementares)
print(f"Dicionário após update: {dicionario_aluno}")

Dicionário original: {'nome': 'Ana', 'idade': 20, 'curso': 'Engenharia', 'notas': [8.5, 9.0, 7.8]}
Nome do aluno: Ana
Idade do aluno (usando get): 20
Curso do aluno (chave inexistente com get, valor padrão): Não informado
Dicionário após modificar idade: {'nome': 'Ana', 'idade': 21, 'curso': 'Engenharia', 'notas': [8.5, 9.0, 7.8]}
Dicionário após adicionar cidade: {'nome': 'Ana', 'idade': 21, 'curso': 'Engenharia', 'notas': [8.5, 9.0, 7.8], 'cidade': 'São Paulo'}
Curso removido: Engenharia
Dicionário após remover curso: {'nome': 'Ana', 'idade': 21, 'notas': [8.5, 9.0, 7.8], 'cidade': 'São Paulo'}
Dicionário após remover notas com del: {'nome': 'Ana', 'idade': 21, 'cidade': 'São Paulo'}
Chaves do dicionário: dict_keys(['nome', 'idade', 'cidade'])
Valores do dicionário: dict_values(['Ana', 21, 'São Paulo'])
Itens do dicionário: dict_items([('nome', 'Ana'), ('idade', 21), ('cidade', 'São Paulo')])
Número de pares no dicionário: 3
Dicionário após update: {'nome': 'Ana', 'idade': 21, 'cidad

 ### Exercício 2: Gerenciamento de Contato

Crie um dicionário chamado `contato` com as seguintes informações iniciais:

*   `nome`: "Carlos Silva"
*   `telefone`: "(11) 98765-4321"
*   `email`: "carlos.silva@example.com"

Em seguida, realize as seguintes operações:

1.  Adicione um novo campo `endereco`: "Rua das Flores, 123 - Centro".
2.  Altere o `telefone` para "(11) 99887-6655".
3.  Remova o campo `email` do dicionário.
4.  Imprima o dicionário final, todas as suas chaves e todos os seus valores.

In [21]:
# Escreva o código aqui
contato = {
    "nome": "Carlos Silva",
    "telefone": "(11) 98765-4321",
    "email": "carlos.silva@example.com"
}
#1
contato["endereco"] = "Rua das Flores, 123 - Centro"
print(contato)
#2
contato['telefone'] = "(11) 99887-6655"
print(contato)
#3 e 4#
contato.pop('email')
print(contato)

{'nome': 'Carlos Silva', 'telefone': '(11) 98765-4321', 'email': 'carlos.silva@example.com', 'endereco': 'Rua das Flores, 123 - Centro'}
{'nome': 'Carlos Silva', 'telefone': '(11) 99887-6655', 'email': 'carlos.silva@example.com', 'endereco': 'Rua das Flores, 123 - Centro'}
{'nome': 'Carlos Silva', 'telefone': '(11) 99887-6655', 'endereco': 'Rua das Flores, 123 - Centro'}


 ## 3. Tuplas

### Teoria Básica

Tuplas em Python são coleções ordenadas e **imutáveis** de itens. Isso significa que, uma vez criada, uma tupla não pode ter seus elementos adicionados, removidos ou modificados. Elas podem armazenar elementos de diferentes tipos de dados. São definidas usando parênteses `()` e os itens são separados por vírgulas. A imutabilidade torna as tuplas ideais para dados que não devem mudar, como coordenadas ou registros fixos.

In [ ]:
# Exemplo de uso de Tuplas

# Criando uma tupla
minha_tupla = (10, 'dez', 20.5, False)
print(f"Tupla original: {minha_tupla}")

# Tupla com um único elemento (note a vírgula)
tupla_unitaria = (5,)
print(f"Tupla unitária: {tupla_unitaria}, tipo: {type(tupla_unitaria)}")

# Acessando elementos (índices começam em 0)
print(f"Primeiro elemento: {minha_tupla[0]}")
print(f"Terceiro elemento: {minha_tupla[2]}")

# Tentativa de modificar um elemento (resultará em erro)
try:
    minha_tupla[1] = 'vinte'
except TypeError as e:
    print(f"Erro ao tentar modificar tupla: {e}")

# Concatenando tuplas (cria uma nova tupla)
tupla_extra = ('a', 'b')
tupla_concatenada = minha_tupla + tupla_extra
print(f"Tupla concatenada: {tupla_concatenada}")

# Verificando o tamanho da tupla
print(f"Tamanho da tupla: {len(minha_tupla)}")

# Contando ocorrências de um elemento
tupla_numeros = (1, 2, 2, 3, 4, 2)
print(f"Número de '2' na tupla: {tupla_numeros.count(2)}")

# Encontrando o índice da primeira ocorrência de um elemento
print(f"Índice da primeira ocorrência de '3': {tupla_numeros.index(3)}")

# Desempacotamento de tuplas
coordenadas = (10.0, 20.0)
x, y = coordenadas
print(f"Coordenada X: {x}, Coordenada Y: {y}")

# Tuplas podem ser usadas como chaves em dicionários (por serem imutáveis)
dicionario_com_tupla_chave = {
    (1, 2): 'Ponto A',
    (3, 4): 'Ponto B'
}
print(f"Valor para a chave (1, 2): {dicionario_com_tupla_chave[(1, 2)]}")

Tupla original: (10, 'dez', 20.5, False)
Tupla unitária: (5,), tipo: <class 'tuple'>
Primeiro elemento: 10
Terceiro elemento: 20.5
Erro ao tentar modificar tupla: 'tuple' object does not support item assignment
Tupla concatenada: (10, 'dez', 20.5, False, 'a', 'b')
Tamanho da tupla: 4
Número de '2' na tupla: 3
Índice da primeira ocorrência de '3': 3
Coordenada X: 10.0, Coordenada Y: 20.0
Valor para a chave (1, 2): Ponto A


 ### Exercício 3: Informações de Livro

Crie uma tupla chamada `informacoes_livro` para armazenar os seguintes dados de um livro:

*   Título: "O Pequeno Príncipe"
*   Autor: "Antoine de Saint-Exupéry"
*   Ano de Publicação: 1943
*   Gênero: "Fábula"

Em seguida:

1.  Acesse e imprima o título e o autor do livro.
2.  Tente alterar o ano de publicação para 1942 e observe o resultado (comentando o código que causaria erro, se necessário).
3.  Imprima o comprimento total da tupla.

In [25]:
# Escreva o código aqui
informacoes_livro = (
    "O Pequeno Príncipe",
    "Antoine de Saint-Exupéry",
    1943,
    "Fábula"
)

#1
print(f"Título: {informacoes_livro[0]}")
print(f"Autor: {informacoes_livro[1]}")

#2
try:
    informacoes_livro[2] = 1942
except TypeError as e:
    print(f"Erro ao tentar modificar tupla: {e}")

#3
print(f"Comprimento da tupla: {len(informacoes_livro)}")

Título: O Pequeno Príncipe
Autor: Antoine de Saint-Exupéry
Erro ao tentar modificar tupla: 'tuple' object does not support item assignment
Comprimento da tupla: 4


 ## 4. NumPy

### Teoria Básica

NumPy (Numerical Python) é uma biblioteca fundamental para computação científica em Python. Ela fornece um objeto de array multidimensional de alto desempenho, e ferramentas para trabalhar com esses arrays. O NumPy é a base para muitas outras bibliotecas científicas em Python, como SciPy, Matplotlib e scikit-learn.

#### Para que usar?

*   **Arrays Multidimensionais (ndarray)**: Permite armazenar e manipular grandes coleções de dados numéricos de forma eficiente.
*   **Funções Matemáticas para Arrays**: Oferece uma vasta coleção de funções matemáticas para operar em arrays inteiros, sem a necessidade de loops explícitos em Python, o que resulta em um desempenho muito superior.
*   **Computação Numérica Eficiente**: Implementado em C e Fortran, o NumPy é significativamente mais rápido que as listas Python para operações numéricas em grandes volumes de dados.
*   **Broadcasting**: Um mecanismo que permite que o NumPy opere em arrays com diferentes formas durante operações aritméticas.
*   **Vetorização**: A capacidade de aplicar operações a um array inteiro de uma vez, em vez de elemento por elemento, melhorando o desempenho e a legibilidade do código.

#### Como instalar?

NumPy geralmente já vem pré-instalado em ambientes como o Google Colab. Caso precise instalá-lo em seu ambiente local, você pode usar o pip:

```bash
pip install numpy
```

#### Como usar?

Para começar a usar o NumPy em seu código, você geralmente o importa com o alias `np`:

```python
import numpy as np
```

In [ ]:
# Importando a biblioteca NumPy
import numpy as np

  ### Arrays Multidimensionais

O objeto central do NumPy é o `ndarray`, que significa "N-dimensional array". É um contêiner homogêneo de elementos (todos os elementos devem ser do mesmo tipo de dado).

*   **Criação de Arrays**: Pode ser feito a partir de listas Python ou usando funções como [`np.zeros`](https://numpy.org/doc/stable/reference/generated/numpy.zeros.html), [`np.ones`](https://numpy.org/doc/stable/reference/generated/numpy.ones.html), [`np.arange`](https://numpy.org/doc/stable/reference/generated/numpy.arange.html), [`np.linspace`](https://numpy.org/doc/stable/reference/generated/numpy.linspace.html), [`np.random.rand`](https://numpy.org/doc/stable/reference/random/generated/numpy.random.rand.html), etc.
*   **Acesso e Indexação**: Similar às listas Python, mas com capacidades de indexação avançadas.
    *   **Slicing (Fatiamento)**: Permite extrair subarrays ou subconjuntos de elementos usando a notação `[start:stop:step]`, similar às listas Python.
    *   **Indexação Booleana**: Utiliza um array booleano (verdadeiro/falso) para selecionar elementos que satisfazem uma determinada condição. A forma do array booleano deve ser a mesma do array original ou broadcasting compatível.
    *   **Fancy Indexing (Indexação Avançada)**: Permite selecionar elementos ou linhas/colunas arbitrárias usando arrays de índices inteiros. Isso cria uma *cópia* dos dados, ao contrário do slicing que retorna uma *view*.
*   **Atributos Importantes**: [`shape`](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.shape.html) (dimensões do array), [`ndim`](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.ndim.html) (número de dimensões), [`size`](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.size.html) (número total de elementos), [`dtype`](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.dtype.html) (tipo de dado dos elementos).


In [ ]:
# Criando um array 1D a partir de uma lista
arr1d = np.array([1, 2, 3, 4, 5])
print(f"Array 1D: {arr1d}")
print(f"Shape do array 1D: {arr1d.shape}") # (5,)
print(f"Dimensões do array 1D: {arr1d.ndim}") # 1

# Criando um array 2D (matriz) a partir de uma lista de listas
arr2d = np.array([[1, 2, 3], [4, 5, 6]])
print(f"\nArray 2D:\n{arr2d}")
print(f"Shape do array 2D: {arr2d.shape}") # (2, 3)
print(f"Dimensões do array 2D: {arr2d.ndim}") # 2

# Criando um array de zeros
zeros_array = np.zeros((3, 4))
print(f"\nArray de zeros (3x4):\n{zeros_array}")

# Criando um array com valores sequenciais
ranged_array = np.arange(0, 10, 2) # Start, Stop (exclusive), Step
print(f"\nArray com arange: {ranged_array}")

# Acessando elementos
print(f"\nPrimeiro elemento do arr1d: {arr1d[0]}")
print(f"Elemento na linha 0, coluna 1 do arr2d: {arr2d[0, 1]}")

# Slicing (fatiamento)
print(f"Primeiras 3 elementos do arr1d: {arr1d[:3]}")
print(f"Elementos do arr1d do índice 1 ao 3: {arr1d[1:4]}")
print(f"Primeira linha do arr2d: {arr2d[0, :]}")
print(f"Segunda coluna do arr2d: {arr2d[:, 1]}")
print(f"Submatriz de arr2d: {arr2d[0:1, 1:3]}")

# Indexação Booleana
arr_bool = np.array([10, 20, 30, 40, 50, 60])
condicao = arr_bool > 30
print(f"\nArray para indexação booleana: {arr_bool}")
print(f"Condição (arr_bool > 30): {condicao}")
print(f"Elementos maiores que 30: {arr_bool[condicao]}")

# Fancy Indexing
arr_fancy = np.array(['a', 'b', 'c', 'd', 'e'])
indices = [0, 2, 4]
print(f"\nArray para fancy indexing: {arr_fancy}")
print(f"Elementos nos índices {indices}: {arr_fancy[indices]}")

matrix_fancy = np.array([[10, 20, 30], [40, 50, 60], [70, 80, 90]])
linhas_selecionadas = np.array([0, 2])
colunas_selecionadas = np.array([1, 0])
print(f"\nMatriz para fancy indexing:\n{matrix_fancy}")
print(f"Elementos específicos (linhas, colunas): {matrix_fancy[linhas_selecionadas, colunas_selecionadas]}") # Retorna (matrix_fancy[0,1], matrix_fancy[2,0])

Array 1D: [1 2 3 4 5]
Shape do array 1D: (5,)
Dimensões do array 1D: 1

Array 2D:
[[1 2 3]
 [4 5 6]]
Shape do array 2D: (2, 3)
Dimensões do array 2D: 2

Array de zeros (3x4):
[[0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]]

Array com arange: [0 2 4 6 8]

Primeiro elemento do arr1d: 1
Elemento na linha 0, coluna 1 do arr2d: 2
Primeiras 3 elementos do arr1d: [1 2 3]
Elementos do arr1d do índice 1 ao 3: [2 3 4]
Primeira linha do arr2d: [1 2 3]
Segunda coluna do arr2d: [2 5]
Submatriz de arr2d: [[2 3]]

Array para indexação booleana: [10 20 30 40 50 60]
Condição (arr_bool > 30): [False False False  True  True  True]
Elementos maiores que 30: [40 50 60]

Array para fancy indexing: ['a' 'b' 'c' 'd' 'e']
Elementos nos índices [0, 2, 4]: ['a' 'c' 'e']

Matriz para fancy indexing:
[[10 20 30]
 [40 50 60]
 [70 80 90]]
Elementos específicos (linhas, colunas): [20 70]


 ### Exercício 4: Criação e Atributos de Array 2D

1. Crie um array NumPy 2D de 4x3 preenchido com números inteiros de 1 a 12.
2. Imprima o array, seu `shape`, `ndim` e `size`.

In [ ]:
# Escreva o código aqui

 ### Exercício 5: Slicing em Arrays 2D

Dada a matriz `matriz = np.array([[10, 20, 30], [40, 50, 60], [70, 80, 90]])`:
1. Extraia a primeira linha.
2. Extraia a segunda coluna.
3. Extraia a submatriz `[[20, 30], [50, 60]]`.

In [ ]:
# Escreva o código aqui

 ### Exercício 6: Indexação Booleana

Crie um array 1D com 10 números inteiros aleatórios entre 1 e 100. Em seguida, use indexação booleana para:
1. Selecionar e imprimir todos os números pares.
2. Selecionar e imprimir todos os números maiores que 70.

In [ ]:
# Escreva o código aqui

 ### Exercício 7: Fancy Indexing

Dada a matriz `data = np.array([['A', 'B', 'C'], ['D', 'E', 'F'], ['G', 'H', 'I']])`:
1. Selecione os elementos nas posições `(0, 1)` (B) e `(2, 0)` (G) usando fancy indexing.
2. Selecione as linhas de índice 0 e 2.

In [ ]:
# Escreva o código aqui

  ### Computação Numérica

O NumPy oferece um desempenho superior para operações matemáticas em arrays em comparação com as listas Python. Ele permite realizar operações elemento a elemento de forma eficiente.

*   **Operações Aritméticas Básicas**: `+`, `-`, `*`, `/`, `**` (potência) aplicadas elemento a elemento.
*   **Funções Universais (ufuncs)**: Funções como [`np.sqrt()`](https://numpy.org/doc/stable/reference/generated/numpy.sqrt.html), [`np.sin()`](https://numpy.org/doc/stable/reference/generated/numpy.sin.html), [`np.log()`](https://numpy.org/doc/stable/reference/generated/numpy.log.html), [`np.exp()`](https://numpy.org/doc/stable/reference/generated/numpy.exp.html) que operam elemento a elemento.
*   **Funções de Agregação**: [`np.sum()`](https://numpy.org/doc/stable/reference/generated/numpy.sum.html), [`np.mean()`](https://numpy.org/doc/stable/reference/generated/numpy.mean.html), [`np.min()`](https://numpy.org/doc/stable/reference/generated/numpy.min.html), [`np.max()`](https://numpy.org/doc/stable/reference/generated/numpy.max.html), [`np.std()`](https://numpy.org/doc/stable/reference/generated/numpy.std.html) para calcular estatísticas sobre arrays inteiros ou ao longo de um eixo específico.


In [ ]:
# Criando dois arrays para demonstração
a = np.array([10, 20, 30, 40])
b = np.array([1, 2, 3, 4])

# Operações aritméticas elemento a elemento
print(f"Soma: {a + b}")
print(f"Subtração: {a - b}")
print(f"Multiplicação: {a * b}")
print(f"Divisão: {a / b}")
print(f"Potência: {a ** 2}") # Eleva cada elemento de 'a' ao quadrado

# Funções universais (ufuncs)
print(f"Raiz quadrada de 'a': {np.sqrt(a)}")
print(f"Seno de 'b': {np.sin(b)}")

# Funções de agregação
matrix = np.array([[1, 2, 3], [4, 5, 6]])
print(f"\nMatrix:\n{matrix}")
print(f"Soma de todos os elementos: {np.sum(matrix)}")
print(f"Média de todos os elementos: {np.mean(matrix):.2f}")
print(f"Soma ao longo do eixo 0 (colunas): {np.sum(matrix, axis=0)}")
print(f"Soma ao longo do eixo 1 (linhas): {np.sum(matrix, axis=1)}")

Soma: [11 22 33 44]
Subtração: [ 9 18 27 36]
Multiplicação: [ 10  40  90 160]
Divisão: [10. 10. 10. 10.]
Potência: [ 100  400  900 1600]
Raiz quadrada de 'a': [3.16227766 4.47213595 5.47722558 6.32455532]
Seno de 'b': [ 0.84147098  0.90929743  0.14112001 -0.7568025 ]

Matrix:
[[1 2 3]
 [4 5 6]]
Soma de todos os elementos: 21
Média de todos os elementos: 3.50
Soma ao longo do eixo 0 (colunas): [5 7 9]
Soma ao longo do eixo 1 (linhas): [ 6 15]


 ### Exercício 8: Operações Aritméticas e Funções Universais

1. Crie um array `x = np.array([1, 2, 3, 4, 5])`.
2. Calcule `x` ao quadrado (`x**2`) e o logaritmo natural de `x` (`np.log(x)`).
3. Some 10 a todos os elementos de `x`.

In [ ]:
# Escreva o código aqui

 ### Exercício 9: Agregação em Arrays 2D

Dada a matriz `m = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])`:
1. Calcule a soma de todos os elementos.
2. Calcule a média de cada coluna.
3. Calcule o valor máximo de cada linha.

In [ ]:
# Escreva o código aqui

 ### Exercício 10: Manipulação de Dados de Temperatura

Você tem as temperaturas diárias em Celsius por uma semana: `temperaturas_c = np.array([22, 24, 21, 23, 25, 20, 22])`.
1. Converta as temperaturas para Fahrenheit usando a fórmula `F = C * 9/5 + 32`.
2. Encontre a temperatura mínima e máxima em Fahrenheit.

In [ ]:
# Escreva o código aqui

 ### Exercício 11: Normalização de Dados

Normalize o array `dados = np.array([10, 20, 30, 40, 50])` para que seus valores fiquem entre 0 e 1. A fórmula é `(x - min(x)) / (max(x) - min(x))`.

In [ ]:
# Escreva o código aqui

 ### Broadcasting

Broadcasting é o termo que descreve como o NumPy trata arrays com diferentes shapes (formas) durante operações aritméticas. Sujeito a certas restrições, o array menor é "transmitido" (broadcast) ao array maior para que eles tenham shapes compatíveis.

As regras de broadcasting são:

1.  Se os arrays não tiverem o mesmo número de dimensões, a forma do array com menos dimensões é prefixada com uns.
2.  Se a forma dos arrays não coincidir em nenhuma dimensão, o array com a forma igual a 1 nessa dimensão é esticado para corresponder à outra forma.
3.  Se em alguma dimensão os tamanhos forem diferentes e nenhum for 1, ocorre um erro (ValueError).

In [ ]:
# Array 1D e escalar
arr = np.array([1, 2, 3])
scalar = 2
print(f"Array: {arr}")
print(f"Escalar: {scalar}")
print(f"Array + Escalar (broadcasting): {arr + scalar}") # 2 é broadcast para [2, 2, 2]

# Array 2D e array 1D
A = np.array([[1, 2, 3], [4, 5, 6]])
b = np.array([10, 20, 30])
print(f"\nMatriz A:\n{A}")
print(f"Vetor b: {b}")
print(f"A + b (broadcasting):\n{A + b}") # b é broadcast para [[10, 20, 30], [10, 20, 30]]

# Array 2D e array 1D (coluna)
c = np.array([[10], [20]])
print(f"\nMatriz A:\n{A}")
print(f"Vetor c (coluna):\n{c}")
print(f"A + c (broadcasting):\n{A + c}") # c é broadcast para [[10, 10, 10], [20, 20, 20]]

Array: [1 2 3]
Escalar: 2
Array + Escalar (broadcasting): [3 4 5]

Matriz A:
[[1 2 3]
 [4 5 6]]
Vetor b: [10 20 30]
A + b (broadcasting):
[[11 22 33]
 [14 25 36]]

Matriz A:
[[1 2 3]
 [4 5 6]]
Vetor c (coluna):
[[10]
 [20]]
A + c (broadcasting):
[[11 12 13]
 [24 25 26]]


 ### Exercício 12: Broadcasting Escalar

Crie um array `vetor = np.array([1, 2, 3, 4, 5])`. Multiplique todos os elementos por 5 usando broadcasting e imprima o resultado.

In [ ]:
# Escreva o código aqui

 ### Exercício 13: Broadcasting de Array 1D com Matriz (Linha)

Crie uma matriz 3x3 `M = np.ones((3, 3))` e um array 1D `v = np.array([1, 2, 3])`. Some `v` a cada *linha* de `M` usando broadcasting e imprima o resultado.

In [ ]:
# Escreva o código aqui

 ### Exercício 14: Broadcasting de Array 1D com Matriz (Coluna)

Crie uma matriz 3x2 `M = np.arange(6).reshape(3, 2)` e um array 1D `v_coluna = np.array([[10], [20], [30]])`. Some `v_coluna` a cada *coluna* de `M` usando broadcasting e imprima o resultado.

In [ ]:
# Escreva o código aqui

 ### Exercício 15: Cenário de Média Ponderada com Broadcasting

Você tem uma matriz de notas de alunos `notas = np.array([[8, 7, 9], [6, 8, 7], [9, 9, 8]])` e um vetor de pesos para cada prova `pesos = np.array([0.3, 0.4, 0.3])`. Calcule a média ponderada de cada aluno usando broadcasting. (Dica: use `np.sum` com o argumento `axis`).

In [ ]:
# Escreva o código aqui

 ### Vetorização

Vetorização é o processo de escrever operações em arrays de forma a eliminar loops explícitos. Em vez de iterar sobre os elementos de um array um por um, as operações são aplicadas a todo o array de uma vez. Isso é crucial para o desempenho no NumPy, pois as operações vetorizadas são implementadas em C e são muito mais rápidas do que os loops em Python.

*   **Benefícios**: Melhor desempenho, código mais limpo e conciso, e menor chance de erros.

In [ ]:
# Exemplo de operação não vetorizada (usando loop Python)
lista = [1, 2, 3, 4, 5]
resultado_loop = []
import time

start_time = time.time()
for x in lista:
    resultado_loop.append(x * 2)
end_time = time.time()
print(f"Resultado com loop: {resultado_loop}")
print(f"Tempo com loop: {end_time - start_time:.6f} segundos")

# Exemplo de operação vetorizada com NumPy
numpy_array = np.array([1, 2, 3, 4, 5])

start_time = time.time()
resultado_vetorizado = numpy_array * 2
end_time = time.time()
print(f"\nResultado vetorizado: {resultado_vetorizado}")
print(f"Tempo vetorizado: {end_time - start_time:.6f} segundos")

# Demonstração em um cenário maior
large_array = np.random.rand(1_000_000)
large_list = large_array.tolist()

# Tempo para loop Python
start_time = time.time()
result_list = []
for x in large_list:
    result_list.append(np.sqrt(x))
end_time = time.time()
print(f"\nTempo para raiz quadrada com loop em lista de 1M elementos: {end_time - start_time:.6f} segundos")

# Tempo para operação vetorizada com NumPy
start_time = time.time()
result_numpy = np.sqrt(large_array)
end_time = time.time()
print(f"Tempo para raiz quadrada vetorizada em array de 1M elementos: {end_time - start_time:.6f} segundos")

# A diferença de desempenho é clara em grandes datasets.

Resultado com loop: [2, 4, 6, 8, 10]
Tempo com loop: 0.000079 segundos

Resultado vetorizado: [ 2  4  6  8 10]
Tempo vetorizado: 0.000067 segundos

Tempo para raiz quadrada com loop em lista de 1M elementos: 1.086648 segundos
Tempo para raiz quadrada vetorizada em array de 1M elementos: 0.004833 segundos


 ### Exercício 16: Comparação de Desempenho (Soma)

1. Crie uma lista Python com 1 milhão de números aleatórios.
2. Crie um array NumPy equivalente com os mesmos números.
3. Calcule a soma de todos os elementos usando um loop `for` para a lista Python e compare o tempo com `np.sum()` para o array NumPy.

In [ ]:
# Escreva o código aqui

 ### Exercício 17: Operação Vetorizada Simples

Crie um array NumPy com 100.000 números inteiros aleatórios entre 1 e 100. Multiplique todos os elementos por 10 e subtraia 5 de cada um, tudo em uma única operação vetorizada. Imprima os primeiros 5 elementos do resultado.

In [ ]:
# Escreva o código aqui

 ### Exercício 18: Operação Vetorizada com Funções Universais

Crie um array NumPy de 500.000 números entre 0 e `2*np.pi` usando `np.linspace`. Calcule o seno de todos esses números usando `np.sin()` em uma única operação vetorizada. Imprima os primeiros 5 valores e o último valor.

In [ ]:
# Escreva o código aqui

 ### Exercício 19: Filtragem e Modificação Vetorizada

Crie um array NumPy de 1 milhão de números aleatórios de ponto flutuante entre 0 e 1.
1. Use uma operação vetorizada para substituir todos os valores menores que 0.2 por 0.
2. Use uma operação vetorizada para substituir todos os valores maiores que 0.8 por 1.
3. Imprima a contagem de zeros e uns no array modificado.

In [ ]:
# Escreva o código aqui

  ## 5. Funções Estatísticas Avançadas com NumPy

### Teoria Básica

Além das funções de agregação básicas (`sum`, `mean`, `min`, `max`), o NumPy oferece uma gama de funções estatísticas mais avançadas para análises de dados. Essas funções operam de forma eficiente em arrays, aproveitando a vetorização para performance.

*   [`np.median()`](https://numpy.org/doc/stable/reference/generated/numpy.median.html): Calcula a mediana de um array ou ao longo de um eixo.
*   [`np.std()`](https://numpy.org/doc/stable/reference/generated/numpy.std.html): Calcula o desvio padrão de um array.
*   [`np.var()`](https://numpy.org/doc/stable/reference/generated/numpy.var.html): Calcula a variância de um array.
*   [`np.percentile(array, q)`](https://numpy.org/doc/stable/reference/generated/numpy.percentile.html): Calcula o q-ésimo percentil dos dados.
*   [`np.corrcoef(x, y)`](https://numpy.org/doc/stable/reference/generated/numpy.corrcoef.html): Calcula o coeficiente de correlação de Pearson entre duas variáveis.


In [ ]:
import numpy as np

# Exemplo de Funções Estatísticas Avançadas

dados = np.array([10, 12, 15, 12, 18, 20, 22, 15, 25, 30])

print(f"Dados: {dados}")
print(f"Mediana: {np.median(dados)}")
print(f"Desvio Padrão: {np.std(dados):.2f}")
print(f"Variância: {np.var(dados):.2f}")
print(f"25º Percentil: {np.percentile(dados, 25)}")
print(f"75º Percentil: {np.percentile(dados, 75)}")

# Exemplo com matriz e eixos
matriz_estatistica = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
print(f"\nMatriz:\n{matriz_estatistica}")
print(f"Mediana ao longo do eixo 0 (colunas): {np.median(matriz_estatistica, axis=0)}")
print(f"Mediana ao longo do eixo 1 (linhas): {np.median(matriz_estatistica, axis=1)}")

# Coeficiente de correlação
x_corr = np.array([1, 2, 3, 4, 5])
y_corr = np.array([2, 4, 5, 4, 5])
print(f"\nCoeficiente de correlação entre x e y:\n{np.corrcoef(x_corr, y_corr)}")

Dados: [10 12 15 12 18 20 22 15 25 30]
Mediana: 16.5
Desvio Padrão: 6.06
Variância: 36.69
25º Percentil: 12.75
75º Percentil: 21.5

Matriz:
[[1 2 3]
 [4 5 6]
 [7 8 9]]
Mediana ao longo do eixo 0 (colunas): [4. 5. 6.]
Mediana ao longo do eixo 1 (linhas): [2. 5. 8.]

Coeficiente de correlação entre x e y:
[[1.         0.77459667]
 [0.77459667 1.        ]]


 ### Exercício 20: Análise Estatística Básica

Dado o array `alturas = np.array([1.72, 1.80, 1.65, 1.75, 1.90, 1.78])`:
1. Calcule e imprima a mediana das alturas.
2. Calcule e imprima o desvio padrão das alturas (arredonde para 2 casas decimais).
3. Calcule e imprima a variância das alturas (arredonde para 2 casas decimais).

In [ ]:
# Escreva o código aqui

 ### Exercício 21: Percentis de Dados de Vendas

Você tem os dados de vendas diárias de um produto em um mês: `vendas = np.array([120, 150, 130, 140, 160, 180, 170, 125, 135, 145, 155, 165, 175, 190, 200, 185, 160, 140, 130, 120, 110, 100, 90, 80, 70, 60, 50, 40, 30, 20])`.
1. Calcule o 10º percentil das vendas.
2. Calcule o 90º percentil das vendas.
3. O que esses percentis indicam?

In [ ]:
# Escreva o código aqui

 ### Exercício 22: Correlação entre Variáveis

Você tem dois arrays representando o número de horas estudadas (`horas_estudo`) e a nota obtida (`nota_final`) por 5 alunos:
`horas_estudo = np.array([2, 3, 5, 6, 8])`
`nota_final = np.array([60, 70, 85, 90, 95])`

Calcule o coeficiente de correlação entre `horas_estudo` e `nota_final` e interprete o resultado.

In [ ]:
# Escreva o código aqui

 ### Exercício 23: Estatísticas por Linha/Coluna em Matriz

Dada a matriz `dados_medicos = np.array([[180, 80, 20], [170, 70, 25], [175, 75, 22]])` onde as colunas representam [Altura (cm), Peso (kg), Idade (anos)]:
1. Calcule a média de cada coluna (Altura, Peso, Idade).
2. Calcule a mediana de cada linha (estatísticas por indivíduo, embora menos comum neste caso).

In [ ]:
# Escreva o código aqui

  ## 6. Sorting, Searching e Counting com NumPy

### Teoria Básica

NumPy oferece funções eficientes para ordenar, buscar e contar elementos em arrays, que são essenciais para análise e manipulação de dados.

*   [`np.sort()`](https://numpy.org/doc/stable/reference/generated/numpy.sort.html): Retorna uma cópia ordenada do array. Pode ordenar por eixos específicos em arrays multidimensionais.
*   [`np.argsort()`](https://numpy.org/doc/stable/reference/generated/numpy.argsort.html): Retorna os *índices* que ordenariam um array. Útil para ordenar múltiplos arrays com base na ordem de um.
*   [`np.where(condition, x, y)`](https://numpy.org/doc/stable/reference/generated/numpy.where.html): Retorna elementos selecionados de `x` ou `y` dependendo da `condition`. Se `x` e `y` não forem fornecidos, retorna os índices onde a condição é verdadeira.
*   [`np.unique()`](https://numpy.org/doc/stable/reference/generated/numpy.unique.html): Retorna os elementos únicos de um array ordenado. Pode retornar também os índices dos elementos únicos, os índices inversos e as contagens.
*   [`np.count_nonzero()`](https://numpy.org/doc/stable/reference/generated/numpy.count_nonzero.html): Conta o número de elementos não-zero em um array.


In [ ]:
import numpy as np

# Exemplo de Sorting, Searching e Counting

arr_dados = np.array([3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5])

print(f"Array original: {arr_dados}")

# Sorting
arr_ordenado = np.sort(arr_dados)
print(f"Array ordenado: {arr_ordenado}")

matriz_sort = np.array([[5, 2, 8], [1, 9, 4]])
print(f"\nMatriz para sort:\n{matriz_sort}")
print(f"Matriz ordenada por linha (axis=1):\n{np.sort(matriz_sort, axis=1)}")
print(f"Matriz ordenada por coluna (axis=0):\n{np.sort(matriz_sort, axis=0)}")

# Argsort
indices_ordenacao = np.argsort(arr_dados)
print(f"\nÍndices que ordenariam o array: {indices_ordenacao}")
print(f"Array ordenado usando argsort: {arr_dados[indices_ordenacao]}")

# Where
indices_maior_que_5 = np.where(arr_dados > 5)
print(f"\nÍndices onde os elementos são maiores que 5: {indices_maior_que_5}")
print(f"Elementos maiores que 5: {arr_dados[indices_maior_que_5]}")

substituicao = np.where(arr_dados > 5, 99, arr_dados)
print(f"Array com elementos maiores que 5 substituídos por 99: {substituicao}")

# Unique
elementos_unicos = np.unique(arr_dados)
print(f"\nElementos únicos: {elementos_unicos}")

elementos_unicos_contagem = np.unique(arr_dados, return_counts=True)
print(f"Elementos únicos e suas contagens: {elementos_unicos_contagem}")

# Count Nonzero
arr_zero = np.array([0, 1, 0, 5, 0, -2])
print(f"\nArray para contar não-zeros: {arr_zero}")
print(f"Número de elementos não-zero: {np.count_nonzero(arr_zero)}")

Array original: [3 1 4 1 5 9 2 6 5 3 5]
Array ordenado: [1 1 2 3 3 4 5 5 5 6 9]

Matriz para sort:
[[5 2 8]
 [1 9 4]]
Matriz ordenada por linha (axis=1):
[[2 5 8]
 [1 4 9]]
Matriz ordenada por coluna (axis=0):
[[1 2 4]
 [5 9 8]]

Índices que ordenariam o array: [ 1  3  6  0  9  2  8  4 10  7  5]
Array ordenado usando argsort: [1 1 2 3 3 4 5 5 5 6 9]

Índices onde os elementos são maiores que 5: (array([5, 7]),)
Elementos maiores que 5: [9 6]
Array com elementos maiores que 5 substituídos por 99: [ 3  1  4  1  5 99  2 99  5  3  5]

Elementos únicos: [1 2 3 4 5 6 9]
Elementos únicos e suas contagens: (array([1, 2, 3, 4, 5, 6, 9]), array([2, 1, 2, 1, 3, 1, 1]))

Array para contar não-zeros: [ 0  1  0  5  0 -2]
Número de elementos não-zero: 3


 ### Exercício 24: Ordenar um Array e Encontrar Únicos

Dado um array de pontuações `pontuacoes = np.array([85, 92, 78, 85, 95, 78, 92, 80])`:
1. Ordene o array em ordem crescente.
2. Encontre e imprima os elementos únicos do array e suas respectivas contagens.

In [ ]:
# Escreva o código aqui

 ### Exercício 25: Seleção Condicional com `np.where`

Crie um array `temperaturas = np.array([25, 18, 30, 22, 15, 28])`. Use `np.where` para criar um novo array onde temperaturas abaixo de 20 sejam substituídas por 0 e as demais permaneçam iguais.

In [ ]:
# Escreva o código aqui

 ### Exercício 26: Ordenação de Múltiplos Arrays (`argsort`)

Você tem idades `idades = np.array([30, 22, 45, 19, 35])` e nomes `nomes = np.array(['João', 'Maria', 'Pedro', 'Ana', 'Carlos'])`.
Use `np.argsort` para ordenar o array `nomes` com base na ordem crescente das `idades`.

In [ ]:
# Escreva o código aqui

 ### Exercício 27: Contar Elementos Não-Zero e Positivos

Dado um array `valores = np.array([-1, 0, 5, -3, 0, 10, 0])`:
1. Conte o número total de elementos não-zero.
2. Conte o número de elementos estritamente positivos (maiores que zero).

In [ ]:
# Escreva o código aqui

  ## 7. Cópia de Arrays (Views vs. Copies) no NumPy

### Teoria Básica

No NumPy, é crucial entender a diferença entre uma 'view' e uma 'copy' ao manipular arrays, pois isso afeta como as modificações são propagadas.

*   **Copy (Cópia)**: Cria um novo array e um novo espaço na memória. Modificações na cópia não afetam o array original, e vice-versa. O método [`copy()`](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.copy.html) explicitamente cria uma cópia.
*   **View (Visão)**: Cria um novo objeto array que *aponta para os mesmos dados* do array original. Modificações na view **afetam** o array original, e vice-versa. Slicing e algumas operações de reshape geralmente retornam views. O método [`view()`](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.view.html) explicitamente cria uma view.

A regra geral é: se você deseja modificar uma parte de um array sem alterar o original, use `.copy()`. Se você quer economizar memória e as modificações devem ser refletidas no original (ou vice-versa), uma view é apropriada.


In [ ]:
import numpy as np

# Exemplo de Views vs. Copies

# Array Original
arr_original = np.array([1, 2, 3, 4, 5])
print(f"Array Original: {arr_original}")

# --- Criando uma View (Slicing) ---
view_array = arr_original[1:4] # Slicing geralmente cria uma view
print(f"View do array (slice): {view_array}")

# Modificando a view
view_array[0] = 99
print(f"View após modificação: {view_array}")
print(f"Array Original após modificação da View: {arr_original}") # O original é alterado!

# Verificando se os dados são os mesmos
print(f"São os mesmos dados? (base da view é o original): {view_array.base is arr_original}")

# --- Criando uma Cópia (Explicitamente com .copy()) ---
arr_original_2 = np.array([10, 20, 30, 40, 50])
copy_array = arr_original_2.copy() # Cria uma cópia explícita
print(f"\nArray Original 2: {arr_original_2}")
print(f"Cópia do array: {copy_array}")

# Modificando a cópia
copy_array[0] = 111
print(f"Cópia após modificação: {copy_array}")
print(f"Array Original 2 após modificação da Cópia: {arr_original_2}") # O original NÃO é alterado!

# Verificando se os dados são os mesmos
print(f"São os mesmos dados? (base da cópia é o original): {copy_array.base is arr_original_2}") # Deve ser False ou None

Array Original: [1 2 3 4 5]
View do array (slice): [2 3 4]
View após modificação: [99  3  4]
Array Original após modificação da View: [ 1 99  3  4  5]
São os mesmos dados? (base da view é o original): True

Array Original 2: [10 20 30 40 50]
Cópia do array: [10 20 30 40 50]
Cópia após modificação: [111  20  30  40  50]
Array Original 2 após modificação da Cópia: [10 20 30 40 50]
São os mesmos dados? (base da cópia é o original): False


 ### Exercício 28: Entendendo Views

1. Crie um array `base_array = np.arange(10)`.
2. Crie uma view `sub_array = base_array[2:7]`.
3. Modifique o terceiro elemento de `sub_array` para 100.
4. Imprima `base_array` e `sub_array` e explique o que aconteceu.

In [ ]:
# Escreva o código aqui

 ### Exercício 29: Criando Cópias Explícitas

1. Crie um array `dados_originais = np.array([10, 20, 30, 40])`.
2. Crie uma cópia explícita deste array chamada `dados_modificados` usando `.copy()`.
3. Altere o primeiro elemento de `dados_modificados` para 5.
4. Imprima ambos os arrays e confirme que `dados_originais` não foi alterado.

In [ ]:
# Escreva o código aqui

 ### Exercício 30: Condições para View ou Copy

Explique em que cenário você optaria por criar uma 'view' de um array em vez de uma 'copy', e vice-versa. Dê um exemplo prático para cada.

In [ ]:
# Escreva o código aqui

 ### Exercício 31: Verificando o 'base' de um Array

1. Crie um array `a = np.array([1, 2, 3])`.
2. Crie `b = a.view()` e `c = a.copy()`.
3. Imprima o `base` de `b` e `c`. O que você observa e por que?

In [ ]:
# Escreva o código aqui

  ## 8. Manipulação de Shape (Reshape, Transpose) no NumPy

### Teoria Básica

A manipulação da forma (shape) dos arrays é uma operação fundamental no NumPy, especialmente quando se trabalha com modelos de Machine Learning que exigem entradas com dimensões específicas. As funções mais comuns são:

*   [`reshape(new_shape)`](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.reshape.html): Retorna um array com o mesmo número de elementos, mas com uma nova forma. O número total de elementos deve ser consistente.
*   [`transpose()`](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.transpose.html) ou [`T`](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.T.html): Inverte ou permuta os eixos de um array. Para matrizes 2D, isso significa trocar linhas por colunas.
*   [`flatten()`](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.flatten.html): Retorna uma cópia do array recolhido em 1D.
*   [`ravel()`](https://numpy.org/doc/stable/reference/generated/numpy.ravel.html): Retorna uma view (se possível) ou uma cópia do array recolhido em 1D. Mais eficiente que `flatten` se uma view for suficiente.
*   [`np.newaxis`](https://numpy.org/doc/stable/reference/constants.html#numpy.newaxis): Usado para aumentar as dimensões de um array, adicionando um novo eixo (por exemplo, para transformar um vetor 1D em um vetor coluna 2D).


In [ ]:
import numpy as np

# Exemplo de Manipulação de Shape

arr_1d = np.arange(12)
print(f"Array 1D: {arr_1d}, Shape: {arr_1d.shape}")

# Reshape para 2D (3 linhas, 4 colunas)
arr_2d_3x4 = arr_1d.reshape(3, 4)
print(f"\nArray 2D (3x4):\n{arr_2d_3x4}, Shape: {arr_2d_3x4.shape}")

# Reshape para 2D (4 linhas, 3 colunas)
arr_2d_4x3 = arr_1d.reshape(4, 3)
print(f"Array 2D (4x3):\n{arr_2d_4x3}, Shape: {arr_2d_4x3.shape}")

# Reshape com -1 (NumPy calcula a dimensão automaticamente)
arr_2d_auto = arr_1d.reshape(2, -1) # 2 linhas, colunas calculadas (6)
print(f"Array 2D (2 linhas, colunas auto):\n{arr_2d_auto}, Shape: {arr_2d_auto.shape}")

# Transpose (para matriz 2D)
matriz_original = np.array([[1, 2, 3], [4, 5, 6]])
print(f"\nMatriz Original:\n{matriz_original}, Shape: {matriz_original.shape}")

matriz_transposta = matriz_original.T # ou matriz_original.transpose()
print(f"Matriz Transposta:\n{matriz_transposta}, Shape: {matriz_transposta.shape}")

# Flatten vs Ravel
flat_copy = matriz_original.flatten()
raveled_view_or_copy = matriz_original.ravel()
print(f"\nArray Original (para flatten/ravel):\n{matriz_original}")
print(f"Flatten (cópia 1D): {flat_copy}, Shape: {flat_copy.shape}")
print(f"Ravel (view/cópia 1D): {raveled_view_or_copy}, Shape: {raveled_view_or_copy.shape}")

# Modificando ravel para ver se é view
raveled_view_or_copy[0] = 999
print(f"Matriz Original após modificar ravel: {matriz_original}") # Original foi alterado, então ravel retornou uma view

# np.newaxis para adicionar dimensão
vetor = np.array([1, 2, 3])
vetor_linha = vetor[np.newaxis, :]
vetor_coluna = vetor[:, np.newaxis]
print(f"\nVetor original: {vetor}, Shape: {vetor.shape}")
print(f"Vetor como linha (1x3): {vetor_linha}, Shape: {vetor_linha.shape}")
print(f"Vetor como coluna (3x1): {vetor_coluna}, Shape: {vetor_coluna.shape}")

Array 1D: [ 0  1  2  3  4  5  6  7  8  9 10 11], Shape: (12,)

Array 2D (3x4):
[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]], Shape: (3, 4)
Array 2D (4x3):
[[ 0  1  2]
 [ 3  4  5]
 [ 6  7  8]
 [ 9 10 11]], Shape: (4, 3)
Array 2D (2 linhas, colunas auto):
[[ 0  1  2  3  4  5]
 [ 6  7  8  9 10 11]], Shape: (2, 6)

Matriz Original:
[[1 2 3]
 [4 5 6]], Shape: (2, 3)
Matriz Transposta:
[[1 4]
 [2 5]
 [3 6]], Shape: (3, 2)

Array Original (para flatten/ravel):
[[1 2 3]
 [4 5 6]]
Flatten (cópia 1D): [1 2 3 4 5 6], Shape: (6,)
Ravel (view/cópia 1D): [1 2 3 4 5 6], Shape: (6,)
Matriz Original após modificar ravel: [[999   2   3]
 [  4   5   6]]

Vetor original: [1 2 3], Shape: (3,)
Vetor como linha (1x3): [[1 2 3]], Shape: (1, 3)
Vetor como coluna (3x1): [[1]
 [2]
 [3]], Shape: (3, 1)


 ### Exercício 32: Reshape para Diferentes Dimensões

1. Crie um array 1D `dados = np.arange(16)`.
2. Redimensione `dados` para uma matriz 4x4 (`4, 4`).
3. Redimensione `dados` para uma matriz 2x8 (`2, 8`).
4. Redimensione `dados` para uma matriz de 8 linhas e o número de colunas calculado automaticamente (`8, -1`).

In [ ]:
# Escreva o código aqui

 ### Exercício 33: Transpor uma Matriz

Crie uma matriz 3x2 `m_original = np.array([[1, 2], [3, 4], [5, 6]])`.
1. Imprima a matriz original e seu shape.
2. Transponha a matriz usando o atributo `.T` e imprima a matriz transposta e seu novo shape.

In [ ]:
# Escreva o código aqui

 ### Exercício 34: Flatten vs Ravel em um Cenário Real

Imagine que você tem uma imagem em escala de cinza representada por uma matriz 5x5 `imagem_cinza = np.random.randint(0, 256, size=(5, 5))`.
1. Use `flatten()` para obter uma versão 1D da imagem (como uma cópia).
2. Use `ravel()` para obter uma versão 1D da imagem. Tente modificar o primeiro elemento da versão `ravel()` e observe se a `imagem_cinza` original é afetada. Por que?

In [ ]:
# Escreva o código aqui

 ### Exercício 35: Adicionar Dimensões com `newaxis`

Você tem um vetor de características `caracteristicas = np.array([10, 20, 30, 40])`.
1. Transforme-o em um vetor coluna (4 linhas, 1 coluna) usando `np.newaxis`.
2. Transforme-o em um vetor linha (1 linha, 4 colunas) usando `np.newaxis`.
3. Imprima os shapes de cada um dos arrays resultantes.

In [ ]:
# Escreva o código aqui

  ## 9. Criação de Arrays Específicos com NumPy

### Teoria Básica

NumPy oferece várias funções convenientes para criar arrays com padrões específicos, além de converter listas Python. Essas funções são úteis para inicializar arrays, criar dados de teste ou para operações matemáticas específicas.

*   [`np.zeros(shape)`](https://numpy.org/doc/stable/reference/generated/numpy.zeros.html): Cria um array preenchido com zeros.
*   [`np.ones(shape)`](https://numpy.org/doc/stable/reference/generated/numpy.ones.html): Cria um array preenchido com uns.
*   [`np.full(shape, fill_value)`](https://numpy.org/doc/stable/reference/generated/numpy.full.html): Cria um array preenchido com um valor específico.
*   [`np.empty(shape)`](https://numpy.org/doc/stable/reference/generated/numpy.empty.html): Cria um array com dados aleatórios (não inicializados), mas com o shape especificado. Pode ser mais rápido que zeros/ones para arrays grandes onde os valores não importam imediatamente.
*   [`np.eye(N)`](https://numpy.org/doc/stable/reference/generated/numpy.eye.html): Cria uma matriz identidade de dimensão `N x N` (diagonal principal com uns, o restante zeros).
*   [`np.diag(v)`](https://numpy.org/doc/stable/reference/generated/numpy.diag.html): Cria uma matriz diagonal com os elementos de `v` na diagonal principal, ou extrai a diagonal de uma matriz.
*   [`np.linspace(start, stop, num)`](https://numpy.org/doc/stable/reference/generated/numpy.linspace.html): Cria um array com `num` elementos igualmente espaçados entre `start` e `stop` (inclusive por padrão).
*   [`np.random.rand(d0, d1, ..., dn)`](https://numpy.org/doc/stable/reference/random/generated/numpy.random.rand.html): Cria um array com valores aleatórios uniformemente distribuídos entre 0 e 1.
*   [`np.random.randint(low, high, size)`](https://numpy.org/doc/stable/reference/random/generated/numpy.random.randint.html): Cria um array com inteiros aleatórios no intervalo `[low, high)`.
*   [`np.fromfunction(function, shape)`](https://numpy.org/doc/stable/reference/generated/numpy.fromfunction.html): Aplica uma função a cada coordenada para gerar valores.


In [ ]:
import numpy as np

# Exemplo de Criação de Arrays Específicos

# Array de zeros (2x3)
zeros = np.zeros((2, 3))
print(f"Array de zeros:\n{zeros}")

# Array de uns (3x2)
ones = np.ones((3, 2))
print(f"\nArray de uns:\n{ones}")

# Array preenchido com um valor específico (2x2 com 7)
full = np.full((2, 2), 7)
print(f"\nArray preenchido com 7:\n{full}")

# Matriz identidade (3x3)
identidade = np.eye(3)
print(f"\nMatriz Identidade:\n{identidade}")

# Matriz diagonal a partir de um vetor
diagonal_vec = np.diag([1, 2, 3])
print(f"\nMatriz Diagonal (a partir de vetor):\n{diagonal_vec}")

# Valores igualmente espaçados (0 a 10, 5 elementos)
linspace = np.linspace(0, 10, 5)
print(f"\nLinspace (0 a 10, 5 elementos): {linspace}")

# Números aleatórios (uniforme 0-1)
random_uniform = np.random.rand(2, 2)
print(f"\nNúmeros aleatórios (uniforme 0-1):\n{random_uniform}")

# Números inteiros aleatórios (1 a 10, 3 elementos)
random_int = np.random.randint(1, 10, size=3)
print(f"Números inteiros aleatórios (1 a 9): {random_int}")

# Criação a partir de função
def func(i, j):
    return i + j
arr_func = np.fromfunction(func, (3, 3), dtype=int)
print(f"\nArray criado a partir de função (i+j):\n{arr_func}")

Array de zeros:
[[0. 0. 0.]
 [0. 0. 0.]]

Array de uns:
[[1. 1.]
 [1. 1.]
 [1. 1.]]

Array preenchido com 7:
[[7 7]
 [7 7]]

Matriz Identidade:
[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]

Matriz Diagonal (a partir de vetor):
[[1 0 0]
 [0 2 0]
 [0 0 3]]

Linspace (0 a 10, 5 elementos): [ 0.   2.5  5.   7.5 10. ]

Números aleatórios (uniforme 0-1):
[[0.41549142 0.73822558]
 [0.73164999 0.82693295]]
Números inteiros aleatórios (1 a 9): [3 2 1]

Array criado a partir de função (i+j):
[[0 1 2]
 [1 2 3]
 [2 3 4]]


 ### Exercício 36: Criação de Matrizes Especiais

1. Crie uma matriz 5x5 preenchida com o valor 10.
2. Crie uma matriz identidade 4x4.
3. Crie uma matriz 3x3 onde a diagonal principal seja `[1, 5, 9]` e o restante zeros.

In [ ]:
# Escreva o código aqui

 ### Exercício 37: Gerando Dados para Gráfico

1. Use `np.linspace` para criar um array `x` com 100 pontos entre -5 e 5.
2. Calcule o `y` correspondente para a função `y = x**2`.
3. Imprima os primeiros 5 e os últimos 5 valores de `x` e `y`.

In [ ]:
# Escreva o código aqui

 ### Exercício 38: Números Aleatórios com Semente

1. Gere um array 3x3 de números inteiros aleatórios entre 1 e 50. Use `np.random.seed(42)` antes de gerar para garantir reprodutibilidade.
2. Imprima o array gerado.
3. Gere outro array 3x3 de números inteiros aleatórios entre 1 e 50 *sem* usar `np.random.seed()` novamente. Observe a diferença.

In [ ]:
# Escreva o código aqui

 ### Exercício 39: Criação com `fromfunction`

Use `np.fromfunction` para criar uma matriz 4x4 onde cada elemento `(i, j)` seja a soma de `i` e `j`.

In [ ]:
# Escreva o código aqui